File: EllipticalRegion.py
#!usr/bin/env python3

import math
from Facet import Template

def prep():
    global sq_colour, x_square, y_square, h, k, a, b, turtle_radius

    # bind fun function
    scene.button.onrelease(play)
    
    turtle.penup()
    # button text
    turtle.setposition(scene.button.xcor(), scene.button.ycor()-10)
    turtle.pendown()
    turtle.write("Move", align="center", font=("Arial",16,"bold"))

    # square parameters
    diameter = 5
    sq_colour = "#ffff00"
    x_square = 50 - scene.wnd_width/2; y_square = 0

    # centre point
    h = 0
    k = 0
    # major/minor axis
    a = scene.wnd_width / 3
    b = scene.wnd_height / 3
    x_dot = h - a; y_dot = k
    # draw ellipse
    turtle.penup()
    turtle.color("red")
    turtle.setposition(x_dot, y_dot)
    turtle.pendown()
    while x_dot < h + a:
        y_dot = k - (b/a)*math.sqrt(math.pow(a, 2) - math.pow((x_dot - h), 2))
        turtle.setposition(x_dot, y_dot)
        y_dot = k + (b/a)*math.sqrt(math.pow(a, 2) - math.pow((x_dot - h), 2))
        turtle.setposition(x_dot, y_dot)
        x_dot += 1
    
    screen.delay(20)
    
    # transform turtle into a square
    turtle.penup()
    turtle.setposition(x_square, y_square)
    turtle.setheading(0)
    turtle.shape("turtle")
    turtle.shapesize(diameter, diameter)
    turtle.color(sq_colour, sq_colour)
    
    turtle_radius = 10*turtle.shapesize()[1]

# fun function when button is clicked
# just moves turtle until it hits the right boundary
def play(x, y):
    global sq_colour, x_square, y_square, h, k, a, b, turtle_radius
    
    if x_square < scene.wnd_width/2 - turtle_radius:

        square_left = x_square - turtle_radius
        square_right = x_square + turtle_radius
        square_top = y_square + turtle_radius
        square_bottom = y_square - turtle_radius

        # determinants for each side of the square
        x_left_det = (b/a)*math.sqrt(abs(math.pow(a, 2) - math.pow((square_left - h), 2)))
        x_right_det = (b/a)*math.sqrt(abs(math.pow(a, 2) - math.pow((square_right - h), 2)))
        y_up_det = (a/b)*math.sqrt(abs(math.pow(b, 2) - math.pow((square_top - k), 2)))
        y_down_det = (a/b)*math.sqrt(abs(math.pow(b, 2) - math.pow((square_bottom - k), 2)))

        # yellow outside the ellipse
        sq_colour = "#ffff00"
        if square_top > k - x_left_det and square_bottom < k + x_left_det and\
           square_top > k - x_right_det and square_bottom < k + x_right_det and\
           square_left > h - y_up_det and square_right < h + y_up_det and\
           square_left > h - y_down_det and square_right < h + y_down_det:
            # green inside the ellipse
            sq_colour = "#00ff00"
        turtle.color(sq_colour, sq_colour)
            
        # change ball's coordinates
        turtle.setposition(x_square, y_square)

        x_square += 10
        screen.ontimer(play(0,0), 500)
    else:
        x_square = 50 - scene.wnd_width/2; y_square = 0
        screen.mainloop()

# Let fly
scene = Template();
turtle = scene.controlButtons()
screen = scene.screen

prep()

File: EllipticalRegion.cpp
#include "stdafx.h"
#include "EllipticalRegion.h"
#include <math.h>


EllipticalRegion::EllipticalRegion(HWND hWnd, int window_width, int window_height)
{
	hWindow = hWnd; // save away window handle
	this->window_width = window_width; // save away window width
	this->window_height = window_height; // save away window width

	sq_colour = RGB(255, 255, 0); // give our square a yellow colour
    // coordinates for the ball(circle)
	x_square = 10;
	y_square = 250;
	previous_x = x_square;
	previous_y = y_square;

	//ellipse coordinates
	h = 450; // vertice
	k = 275; // vertice
	x = h - a;
	y = k - b;

	// Pen and brush matching background colour
	background_brush = CreateSolidBrush(RGB(192, 192, 192));

	// pen(purple colour) for circle
	circle_pen = CreatePen(PS_SOLID, 5, RGB(255, 0, 255));

	// brush for travelling square
	sq_brush = CreateSolidBrush(sq_colour);

	hdc = GetDC(hWindow);
}

/*
* draws the ball/circle using the apt color
*/
void EllipticalRegion::paint() {
	SelectObject(hdc, circle_pen); // use purple colour
	SelectObject(hdc, background_brush); // use background colour
	// draw an ellipse
	Ellipse(hdc, x, y, h + a, k + b);

	// erase previous square
	RECT prev_sq;
	prev_sq.left = previous_x;
	prev_sq.top = previous_y;
	prev_sq.right = previous_x + sqLENGTH;
	prev_sq.bottom = previous_y + sqLENGTH;
	FillRect(hdc, &prev_sq, background_brush);


	// draw square
	RECT sq;
	sq.left = x_square;
	sq.top = y_square;
	sq.right = x_square + sqLENGTH;
	sq.bottom = y_square + sqLENGTH;
	FillRect(hdc, &sq, sq_brush);

	previous_x = x_square;
	previous_y = y_square;
}

/*
* Repeatedly draws square so as to simulate a continuous motion
*/
void EllipticalRegion::ellipsedSquare() {
	// condition for continuing motion
	while (x_square + sqLENGTH < window_width) {
		int square_left = x_square;
		int square_right = x_square + sqLENGTH;
		int square_top = y_square;
		int square_bottom = y_square + sqLENGTH;

		// determinants for each side of the square
		int x_left_det = (int)round(((double)b / a)
			* sqrt(pow(a, 2) - pow((square_left - h), 2)));
		int x_right_det = (int)round(((double)b / a)
			* sqrt(pow(a, 2) - pow((square_right - h), 2)));
		int y_up_det = (int)round(((double)a / b)
			* sqrt(pow(b, 2) - pow((square_top - k), 2)));
		int y_down_det = (int)round(((double)a / b)
			* sqrt(pow(b, 2) - pow((square_bottom - k), 2)));

		if (square_top > k - x_left_det && square_bottom < k + x_left_det
			&& square_top > k - x_right_det && square_bottom < k + x_right_det
			&& square_left > h - y_up_det && square_right < h + y_up_det
			&& square_left > h - y_down_det && square_right < h + y_down_det) {
			// green colour for our moving body(square) inside our ellipse
			sq_colour = RGB(0, 255, 0);
		}
		else {
			// yellow colour for our moving body(square) outside our ellipse
			sq_colour = RGB(255, 255, 0);
		}
		DeleteObject(sq_brush); // delete former brush
		sq_brush = CreateSolidBrush(sq_colour); // recreate brush with a new colour

		paint();

		x_square += 10;
		// introduce a delay betweeen renderings
		Sleep(50);
	}
}


EllipticalRegion::~EllipticalRegion()
{
	DeleteObject(background_brush);
	DeleteObject(circle_pen);
	DeleteObject(sq_brush);
	
	ReleaseDC(hWindow, hdc);
}
